extensions/attributegen/plugin.cc

```diff
@@ -0,0 +1,147 @@
+
+#include "extensions/common/wasm/null/null.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+namespace Null {
+namespace Plugin {
+
+#endif  // NULL_PLUGIN
+
+// END WASM_PROLOG
+
+namespace AttributeGen {
+
+bool PluginRootContext::onConfigure(size_t) {
+  std::unique_ptr<WasmData> configuration = getConfiguration();
+  // Parse configuration JSON string.
+  JsonParseOptions json_options;
+  json_options.ignore_unknown_fields = true;
+  istio::attributegen::PluginConfig config;
+  Status status =
+      JsonStringToMessage(configuration->toString(), &config, json_options);
+  if (status != Status::OK) {
+    LOG_WARN(absl::StrCat(
+        "Cannot parse plugin configuration JSON string [YAML not supported]: ",
+        configuration->toString()));
+    return false;
+  }
+
+  return initAttributeGen(config);
+}
+
+bool PluginRootContext::initAttributeGen(
+    const istio::attributegen::PluginConfig& config) {
+  // Clear prior host resources.
+  // TODO(mjog) see if this still needed since onDone is always called.
+  cleanupAttributeGen();
+
+  for (auto agconfig : config.attributes()) {
+    EvalPhase phase = on_log;
+    if (agconfig.phase() == istio::attributegen::ON_REQUEST) {
+      phase = on_request;
+    }
+    auto ag = AttributeGenerator(phase, agconfig.output_attribute());
+    for (auto matchconfig : agconfig.match()) {
+      uint32_t token = 0;
+      if (matchconfig.condition().empty()) {
+        ag.add_match(Match("", 0, matchconfig.value()));
+        continue;
+      }
+      if (createExpression(matchconfig.condition(), &token) != WasmResult::Ok) {
+        LOG_WARN(absl::StrCat("Cannot create expression: <",
+                              matchconfig.condition(), "> for ",
+                              agconfig.output_attribute()));
+        return false;
+      }
+      LOG_DEBUG(absl::StrCat("Added ", agconfig.output_attribute(), " if (",
+                             matchconfig.condition(), ") -> ",
+                             matchconfig.value()));
+      ag.add_match(Match(matchconfig.condition(), token, matchconfig.value()));
+    }
+    gen_.push_back(ag);
+  }
+  return true;
+}
+
+void PluginRootContext::cleanupAttributeGen() {
+  for (auto ag : gen_) {
+    ag.cleanup();
+  }
+  gen_.clear();
+}
+
+// attributeGen is called on the data path.
+void PluginRootContext::attributeGen(EvalPhase phase) {
+  for (const auto& ag : gen_) {
+    if (phase != ag.phase()) {
+      continue;
+    }
+
+    std::string val;
+    auto eval_status = ag.evaluate(&val);
+    if (!eval_status) {
+      LOG_DEBUG(absl::StrCat("Failed eval. setting ", ag.error_attribute()));
+      // eval failed set error attribute
+      setFilterState(ag.error_attribute(), "1");
+      continue;
+    }
+
+    if (!eval_status.value()) {
+      continue;
+    }
+    LOG_DEBUG(absl::StrCat("Setting ", ag.output_attribute(), " --> ", val));
+    setFilterState(ag.output_attribute(), val);
```

Comment:
I think default span for filterstate is "filterchain", for this plugin, shouldn't it be longer "downstream connection"?
(Not sure if it's available in ABI to do that yet.. would need access to WasmStatePrototype I think)

Reply:
Good question, @kyessenov  @jplevyak ?

Reply:
Yes, it's per filter chain (individual pass per request) if you don't declare the property. The ABI is available, see #2817 .